core/src/workflow/workflow_tasks/concurrency_manager.rs

```diff
@@ -266,15 +266,35 @@ impl WorkflowConcurrencyManager {
         F: for<'a> FnOnce(&'a mut WorkflowManager) -> BoxFuture<Result<Fout>>,
         Fout: Send + Debug,
     {
-        let readlock = self.runs.read();
-        let m = readlock
-            .get(run_id)
-            .ok_or_else(|| WFMachinesError::Fatal("Missing workflow machines".to_string()))?;
-        // This holds a non-async mutex across an await point which is technically a no-no, but
-        // we never access the machines for the same run simultaneously anyway. This should all
-        // get fixed with a generally different approach which moves the runs inside workers.
-        let mut wfm_mutex = m.wfm.lock();
-        let res = mutator(&mut wfm_mutex).await;
+        // TODO: This is less than ideal. We must avoid holding the read lock on the overall machine
+        //  map while async-ly mutating the inner machine. So, we take it out of the map and put
+        //  it back when we're done. We should restructure things to avoid the top-level lock
+        //  on the map.
+
+        let mut wfm = {
+            let readlock = self.runs.read();
+            let m = readlock
+                .get(run_id)
+                .ok_or_else(|| WFMachinesError::Fatal("Missing workflow machines".to_string()))?;
+            let x = m
+                .wfm
+                .lock()
+                .take()
+                .expect("WorkflowManager inside a ManagedRun is never None");
+            x
+        };
+
+        let res = mutator(&mut wfm).await;
+
+        {
+            let readlock = self.runs.read();
+            // SAFETY: Since runs are accessed serially the machines cannot go missing while
+            // we are modifying them for a given run
+            let m = readlock
+                .get(run_id)
+                .expect("Machine cannot be missing as each run is accessed serially");
+            *m.wfm.lock() = Some(wfm);
```

Comment:
Can you hold the option obtained from `lock` above and just use this setter or `replace` or whatever here instead of another map lookup?